File: src/ShipUpdater.ts
import type NodeMWP from "./NodeMWP.js";
import * as Logger from "./Logger.js";
import * as WikiParser from "./WikiParser.js";
import * as Diff from "./Diff.js";
import type { ShipData } from "./interfaces/ShipData.js";
import type { SingleShipData } from "./interfaces/SingleShipData.js";
import APIManager from "./APIManager.js";
import GlobalConfig from "./assets/GlobalConfig.json" with { type: "json" };

/**
 * Ship Updater
 *
 * This is the class that is responsible for facilitating the updating of ship pages.
 */
export default class ShipUpdater {
    BOT: NodeMWP;

    constructor(bot: NodeMWP) {
        this.BOT = bot;
    }

    /**
     * Update all ships with the provided data.
     * 
     * Error handling is performed here
     * @param data The data to update the ships with. If not provided, the data will be fetched from the API
     */
    public async updateAll(data?: ShipData): Promise<void> {
        if (!data) data = await new APIManager().getShipsData();

        for (const SHIP of Object.keys(data)) {
            try {
                await this.updateShip(SHIP, data[SHIP]);
            } catch (error) {
                Logger.log(
                    `Failed to update ship: ${this.getShipName(SHIP)}\n${(error as Error).stack ?? (error as Error).message}`,
                    Logger.LogLevel.ERROR,
                );
            }
        }

        Logger.log(
            "Finished updating all ships",
            Logger.LogLevel.INFO,
            Logger.LogStyle.CHECKMARK,
        );
    }

    /**
     * Get the ship name from the ship data, preforming any necessary name mapping.
     * @param ship The ship to get the name of. Can be a string or a SingleShipData object
     * @private
     */
    private getShipName(ship: SingleShipData | string): string {
        // @ts-expect-error Compiler yells for some reason. We know this is safe (I hope)
        const MAPPED_NAME = GlobalConfig.ship_name_map[
            typeof ship === "string" ? ship : ship.title
        ] as string | undefined;
        return MAPPED_NAME ?? (typeof ship === "string" ? ship : ship.title);
    }

    /**
     * Update an individual ship
     *
     * Error handling is not done here. It is up to the caller to handle errors.
     * @param ship The ship to update
     * @param data The data to update the ship with. If not provided, the data will be fetched from the API
     * @returns
     */
    public async updateShip(
        ship: string,
        data?: SingleShipData,
    ): Promise<void> {
        const UPDATE_START = performance.now();
        //region Data fetching logic
        if (!data) {
            const APIMANAGER = new APIManager();

            const SHIPSTATS: ShipData = await APIMANAGER.getShipsData();
            const SHIPDATA = Object.keys(SHIPSTATS).find(
                (shipData) => shipData === ship,
            );
            if (!SHIPDATA) {
                Logger.log(
                    `Failed to find ship: ${ship}`,
                    Logger.LogLevel.ERROR,
                );
                return;
            }
            data = SHIPSTATS[SHIPDATA];
        }
        //endregion
        ship = this.getShipName(data);
        Logger.log(
            `Updating ship: ${ship}`,
            Logger.LogLevel.INFO,
            Logger.LogStyle.PROGRESS,
        );
        
        //region Article fetch logic
        const FETCH_ARTICLE_START = performance.now();
        const ARTICLE: string = await this.BOT.getArticle(ship) as string;
        if (!ARTICLE) throw new Error(`Failed to fetch article: ${ship}`);
        const FETCH_ARTICLE_END = performance.now();
        Logger.log(
            `Fetching article took ${(FETCH_ARTICLE_END - FETCH_ARTICLE_START).toFixed(2)}ms`,
            Logger.LogLevel.DEBUG,
            Logger.LogStyle.CHECKMARK,
        );
        //endregion

        //region Ignore flag check
        if (
            ARTICLE.toLowerCase().match(/<!--\s*ketchupbot-ignore\s*-->/gim) !==
            null
        )
            throw new Error("Found ignore flag on the page. Skipping update");
        //endregion

        //region Infobox parsing logic
        const INFOBOXPARSESTART = performance.now();
        const PARSED_INFOBOX = WikiParser.parseInfobox(
            WikiParser.extractInfobox(ARTICLE),
        );
        const INFOBOXPARSEEND = performance.now();
        Logger.log(
            `Parsing infobox took ${(INFOBOXPARSEEND - INFOBOXPARSESTART).toFixed(2)}ms`,
            Logger.LogLevel.DEBUG,
            Logger.LogStyle.CHECKMARK,
        );
        //endregion

        //region Convert data into a Partial<Record<string, string>>
        const DATA: Partial<Record<string, string>> = {};
        for (const [KEY, VALUE] of Object.entries(data)) {
            DATA[KEY] = (VALUE as string).toString();
        }
        //endregion

        //region Data merging logic
        const INFOBOX_MERGE_START = performance.now();
        const [MERGED_INFOBOX, UPDATED_PARAMETERS] = WikiParser.mergeData(
            PARSED_INFOBOX,
            DATA,
        );
        const INFOBOX_MERGE_END = performance.now();
        Logger.log(
            `Merging data took ${(INFOBOX_MERGE_END - INFOBOX_MERGE_START).toFixed(2)}ms`,
            Logger.LogLevel.DEBUG,
            Logger.LogStyle.CHECKMARK,
        );
        //endregion

        //region Sanitization logic
        const SANITIZATION_START = performance.now();
        const [SANITIZED_INFOBOX, REMOVED_PARAMETERS] =
            WikiParser.sanitizeData(MERGED_INFOBOX, PARSED_INFOBOX);
        const SANITIZATION_END = performance.now();
        Logger.log(
            `Sanitizing data took ${(SANITIZATION_END - SANITIZATION_START).toFixed(2)}ms`,
            Logger.LogLevel.DEBUG,
            Logger.LogStyle.CHECKMARK,
        );
        //endregion

        //region Diffing logic
        if (JSON.stringify(PARSED_INFOBOX) === JSON.stringify(SANITIZED_INFOBOX)) {
            Logger.log(
                `No changes detected for ship: ${ship}`,
                Logger.LogLevel.INFO,
                Logger.LogStyle.CHECKMARK,
            );
            return;
        }

        if (process.env.NODE_ENV !== "production") console.log(Diff.diffData(PARSED_INFOBOX, SANITIZED_INFOBOX));
        //endregion

        //region Wikitext construction logic
        const WIKITEXT_CONSTRUCTION_START = performance.now();
        const NEW_WIKITEXT = WikiParser.replaceInfobox(
            ARTICLE,
            WikiParser.objectToWikitext(SANITIZED_INFOBOX),
        );
        const WIKITEXT_CONSTRUCTION_END = performance.now();
        Logger.log(
            `Constructing new wikitext took ${(WIKITEXT_CONSTRUCTION_END - WIKITEXT_CONSTRUCTION_START).toFixed(2)}ms`,
            Logger.LogLevel.DEBUG,
            Logger.LogStyle.CHECKMARK,
        );
        //endregion

        //region Article editing logic
        await this.BOT.edit(
            ship,
            NEW_WIKITEXT,
            "Automatic Update" +
                (REMOVED_PARAMETERS.length > 0
                    ? ` | Removed parameters: ${REMOVED_PARAMETERS.sort().join(", ")}`
                    : "") +
                (UPDATED_PARAMETERS.length > 0
                    ? ` | Updated parameters: ${UPDATED_PARAMETERS.filter((param) => !REMOVED_PARAMETERS.includes(param)).sort().join(", ")}`
                    : ""),
        );
        const UPDATEEND = performance.now();
        Logger.log(
            `Updated ship: ${ship} in ${(UPDATEEND - UPDATE_START).toFixed(2)}ms`,
            Logger.LogLevel.INFO,
            Logger.LogStyle.CHECKMARK,
        );
        //endregion
    }
}

File: ketchupbot-updater/ShipUpdater.cs
using System.Diagnostics;
using System.Text.RegularExpressions;
using ketchupbot_updater.API;
using ketchupbot_updater.Types;
using Newtonsoft.Json;

namespace ketchupbot_updater;

/// <summary>
/// Ship updater class to facilitate updating ship pages. You should pass this class to other classes via dependency injection.
/// </summary>
/// <param name="bot"></param>
/// <param name="apiManager"></param>
public partial class ShipUpdater(MwClient bot, ApiManager apiManager)
{
    private static string GetShipName(string data) => GlobalConfiguration.ShipNameMap.GetValueOrDefault(data, data);

    private const int MaxLength = 12;

    /// <summary>
    /// Mass update all ships with the provided data. If no data is provided, it will fetch the data from the API.
    /// </summary>
    /// <param name="shipDatas">The ship data to use during the update run</param>
    /// <param name="multithreaded">Whether to distribute the ship updates over a ThreadPool, or use a singlethreaded foreach loop</param>
    public async Task UpdateAllShips(Dictionary<string, ShipData>? shipDatas = null, bool multithreaded = true)
    {
        var massUpdateStart = Stopwatch.StartNew();
        shipDatas ??= await apiManager.GetShipsData();
        ArgumentNullException.ThrowIfNull(shipDatas);

        if (multithreaded)
        {
            List<Task> tasks = shipDatas.Select(ship => UpdateShipWrapper(ship.Key, ship.Value)).ToList();

            await Task.WhenAll(tasks);
        }
        else
        {
            foreach (KeyValuePair<string, ShipData> ship in shipDatas)
            {
                await UpdateShipWrapper(ship.Key, ship.Value);
            }
        }

        massUpdateStart.Stop();
        Logger.Log($"Finished updating all ships in {massUpdateStart.ElapsedMilliseconds/1000} seconds", style: LogStyle.Checkmark);
        Logger.Log($"Current thread pool count: {ThreadPool.ThreadCount}");
    }

    /// <summary>
    /// Wrapper for UpdateShip that catches exceptions and logs them instead of throwing them.
    /// </summary>
    /// <param name="ship"></param>
    /// <param name="data"></param>
    private async Task UpdateShipWrapper(string ship, ShipData? data = null)
    {
        try
        {
            await UpdateShip(ship, data);
        }
        catch (Exception e)
        {
            Logger.Log($"{GetShipIdentifier(ship)} Failed to update ship: {e.Message}", level: LogLevel.Error);
        }
    }

    /// <summary>
    /// Update a singular ship page with the provided data (or fetch it if not provided)
    /// </summary>
    /// <param name="ship">The name of the ship to update</param>
    /// <param name="data">Supply a <see cref="ShipData"/> object to use for updating. If left null, it will be fetched for you, but this is very bandwidth intensive for mass updating. It is better to grab it beforehand, filter the data for the specific <see cref="ShipData"/> needed, and pass that to the functions.</param>
    public async Task UpdateShip(string ship, ShipData? data = null)
    {
        var updateStart = Stopwatch.StartNew();
        ship = GetShipName(ship);

        #region Data Fetching Logic
        if (data == null)
        {
            Dictionary<string, ShipData>? shipStats = await apiManager.GetShipsData();

            ShipData? shipData = (shipStats ?? throw new InvalidOperationException("Failed to get ship data")).GetValueOrDefault(ship ?? throw new InvalidOperationException("No ship name provided"));

            if (shipData == null)
            {
                Console.WriteLine("Ship not found in API data: " + ship);
                return;
            }

            data = shipData;
        }
        #endregion

        Logger.Log($"{GetShipIdentifier(ship)} Updating ship...", style: LogStyle.Progress);

        #region Article Fetch Logic
#if DEBUG
        var fetchArticleStart = Stopwatch.StartNew();
#endif

        string article = await bot.GetArticle(ship); // Throws exception if article does not exist

#if DEBUG
        fetchArticleStart.Stop();
        Logger.Log($"{GetShipIdentifier(ship)} Fetched article in {fetchArticleStart.ElapsedMilliseconds}ms", style: LogStyle.Checkmark);
#endif
        #endregion

        if (IGNORE_FLAG_REGEX().IsMatch(article.ToLower())) throw new Exception("Found ignore flag in article");

        // I think I can probably come up with a better way to do this. The current way of converting the data to a
        // dictionary and then back to a dictionary is a bit silly. I can probably just use the data object directly,
        // and merge the two objects together somehow. This way we don't have to resort to using Dictionary<string,
        // string> for everything and can use the actual data object which is more strongly typed. The hard part is
        // going to be figuring out how to merge the two objects together. There isn't really a way to enumerate over
        // the properties of an object in C# without using reflection, which is slow. If I went that route, I could
        // probably use reflection to get the properties of the object and then use that to merge the two objects
        // together, but it'll probably have a performance impact. I'll have to think about this a bit more. Also,
        // probably a good idea to read the comment in WikiParser.cs about this

        #region Infobox Parsing Logic
#if DEBUG
        var parsingInfoboxStart = Stopwatch.StartNew();
#endif

        Dictionary<string, string> parsedInfobox = WikiParser.ParseInfobox(WikiParser.ExtractInfobox(article));

#if DEBUG
        parsingInfoboxStart.Stop();
        Logger.Log($"{GetShipIdentifier(ship)} Parsed infobox in {parsingInfoboxStart.ElapsedMilliseconds}ms", style: LogStyle.Checkmark);
#endif
        #endregion

        // Convert data into a dictionary
        var dataDict = JsonConvert.DeserializeObject<Dictionary<string, string>>(JsonConvert.SerializeObject(data));

        #region Data merging logic
#if DEBUG
        var mergeDataStart = Stopwatch.StartNew();
#endif

        Tuple<Dictionary<string, string>, List<string>> mergedData = WikiParser.MergeData(dataDict ?? throw new InvalidOperationException("Supplied data is null after deserialization"), parsedInfobox);

#if DEBUG
        mergeDataStart.Stop();
        Logger.Log($"{GetShipIdentifier(ship)} Merged data in {mergeDataStart.ElapsedMilliseconds}ms", style: LogStyle.Checkmark);
        // JsonSerializerSettings settings = new()
        // {
        //     NullValueHandling = NullValueHandling.Ignore
        // };
        // Console.WriteLine(JsonConvert.SerializeObject(mergedData.Item1, Formatting.Indented, settings));
#endif
        #endregion

        #region Data Sanitization Logic
#if DEBUG
        var sanitizeDataStart = Stopwatch.StartNew();
#endif

        Tuple<Dictionary<string, string>, List<string>> sanitizedData = WikiParser.SanitizeData(mergedData.Item1, parsedInfobox);

#if DEBUG
        sanitizeDataStart.Stop();
        Logger.Log($"{GetShipIdentifier(ship)} Sanitized data in {sanitizeDataStart.ElapsedMilliseconds}ms", style: LogStyle.Checkmark);
#endif
        #endregion

        #region Diffing logic
        // This logic is only for debugging/development instances to see what changes are being made to the infobox. It is not necessary for the bot to function, so it should not be in production.
        // I've turned it off cuz its kinda annoying
#if DEBUG
        // var jdp = new JsonDiffPatch();
        // string? diff = jdp.Diff(JsonConvert.SerializeObject(sanitizedData.Item1, Formatting.Indented), JsonConvert.SerializeObject(parsedInfobox, Formatting.Indented));

        // if (!string.IsNullOrEmpty(diff))
        //     Console.WriteLine($"Diff:\n{diff}");
#endif
        #endregion

        #region Wikitext Construction Logic
#if DEBUG
        var wikitextConstructionStart = Stopwatch.StartNew();
#endif

        string newWikitext = WikiParser.ReplaceInfobox(article, WikiParser.ObjectToWikitext(sanitizedData.Item1));

#if DEBUG
        wikitextConstructionStart.Stop();
        Logger.Log($"{GetShipIdentifier(ship)} Constructed wikitext in {wikitextConstructionStart.ElapsedMilliseconds}ms", style: LogStyle.Checkmark);
#endif
        #endregion

        #region Article Editing Logic
#if DEBUG
        var articleEditStart = Stopwatch.StartNew();
#endif

        await bot.EditArticle(ship, newWikitext, "Automated ship data update");

#if DEBUG
        articleEditStart.Stop();
        Logger.Log($"{GetShipIdentifier(ship)} Edited page in {articleEditStart.ElapsedMilliseconds}ms", style: LogStyle.Checkmark);
#endif
        #endregion

#if DEBUG
        updateStart.Stop();
        Logger.Log($"{GetShipIdentifier(ship)} Updated ship in {updateStart.ElapsedMilliseconds}ms", style: LogStyle.Checkmark);
        Logger.Log($"Current thread pool count: {ThreadPool.ThreadCount}");
#endif
    }

    /// <summary>
    /// Used to get a ship identifier for logging purposes. I'm pretty sure this only runs in debug mode, so it should be okay to have the overhead from the string formatting.
    /// </summary>
    /// <param name="ship"></param>
    /// <returns></returns>
    private static string GetShipIdentifier(string ship)
    {
        string truncatedShipName = ship.Length > MaxLength ? ship[..MaxLength] : ship;
        string paddedShipName = truncatedShipName.PadRight(MaxLength);

        return $"{paddedShipName} {Environment.CurrentManagedThreadId,-2} |";
    }

    [GeneratedRegex(@"<!--\s*ketchupbot-ignore\s*-->", RegexOptions.IgnoreCase | RegexOptions.Multiline)]
    private static partial Regex IGNORE_FLAG_REGEX();
}